ledger: shutdown closes DB while replayBufferedHeadersAsync is iterating, panics with DB Closed

## Summary

A clean SIGTERM during catch-up sync can panic with `DB Closed` if `ledger.replayBufferedHeadersAsync` is mid fork-resolution probe when shutdown phase 3 closes the database. The panic propagates from a worker goroutine and kills the process before deferred cleanup runs (so for example a `--cpuprofile` capture is left empty).

Observed during a profiling run on bb52f1a7 against preview. Stack:

```
panic: DB Closed [recovered, repanicked]
github.com/dgraph-io/badger/v4.(*Txn).NewIterator
github.com/blinklabs-io/dingo/database/plugin/blob/badger.(*BlobStoreBadger).NewIterator
        database/plugin/blob/badger/database.go:515
github.com/blinklabs-io/dingo/database.BlockByHashTxn
        database/block.go:316
github.com/blinklabs-io/dingo/database.BlockByHash
        database/block.go:208
github.com/blinklabs-io/dingo/ledger.(*LedgerState).findPeerForkPath
        ledger/chainsync.go:670
github.com/blinklabs-io/dingo/ledger.(*LedgerState).tryResolveFork
        ledger/chainsync.go:1687
github.com/blinklabs-io/dingo/ledger.(*LedgerState).handleEventChainsyncBlockHeader
        ledger/chainsync.go:1488
github.com/blinklabs-io/dingo/ledger.(*LedgerState).replayBufferedHeaderEvents
        ledger/chainsync.go:950
github.com/blinklabs-io/dingo/ledger.(*LedgerState).replayBufferedHeadersAsync.func1
        ledger/chainsync.go:926
created by ledger.(*LedgerState).replayBufferedHeadersAsync
        ledger/chainsync.go:919
```

Shutdown ordering observed in the log:

```
shutdown phase 1: stopping new work
shutdown phase 1 complete  (elapsed 0)
shutdown phase 2: draining connections
shutdown phase 2 complete  (elapsed 1ms)
shutdown phase 3: flushing state
panic: DB Closed
```

## Why this matters

- Profiling and ops tooling that relies on the `--cpuprofile` / `--memprofile` flags loses captured data because deferred Stop calls in main do not run when a worker goroutine panics.
- A clean operator-initiated shutdown produces a non-zero exit and a panic stack in logs, which looks like data loss.
- The `findPeerForkPath` -> `BlockByHash` -> Badger `NewIterator` path stays hot during catch-up (see #2105), so this race is reachable in practice, not just in profiling.

## Suspected fix

Phase 1 stops new work but the async header-replay goroutine that began before phase 1 keeps running. Phase 3 closes the DB while that goroutine is still issuing reads. Either:

1. Wait for the async header-replay goroutine to drain in phase 2 (treat it as a connection-equivalent worker that must complete before state flush), or
2. Have `replayBufferedHeadersAsync` honor a shutdown context so it stops issuing new DB reads as soon as phase 1 fires, then phase 3 can safely close.

Option 2 is the smaller change and matches how the rest of the connection workers behave.

## Reproduce

1. Load `database/immutable/testdata` into a data dir.
2. Start dingo in serve mode against preview using that data dir.
3. Wait until `chainsync rollback exceeds security parameter K` or any fork-resolution log line shows up.
4. SIGTERM. Observe the `DB Closed` panic in logs and a non-zero exit.

The race does not always reproduce. A second 60 s profiling run with the same setup shut down cleanly. The race is real; just not 100%.

## Notes

- Caught while running for #1893 (operational profiling).
- Related path is the fork-resolution iterator hot path tracked in #2105.

--- a/ledger/chainsync.go
+++ b/ledger/chainsync.go
@@ -935,9 +935,23 @@ func (ls *LedgerState) nextBufferedHeaderConnId() (
 func (ls *LedgerState) replayBufferedHeadersAsync(
 	connId ouroboros.ConnectionId,
 ) {
+	// Bail out if Close has already started so we never spawn a worker
+	// that issues DB reads after phase 3 closes the database (#2107).
+	if ls.closed.Load() {
+		return
+	}
+	ls.replayWG.Add(1)
 	go func() {
+		defer ls.replayWG.Done()
 		ls.chainsyncMutex.Lock()
 		defer ls.chainsyncMutex.Unlock()
+		// Re-check after acquiring the mutex in case Close started
+		// while we were waiting for the lock; the DB reads inside
+		// handleEventChainsyncBlockHeader (BlockByHash etc.) will
+		// panic with "DB Closed" once the owner closes the DB.
+		if ls.closed.Load() {
+			return
+		}
 		if ls.headerPipelineConnId != (ouroboros.ConnectionId{}) ||
 			ls.chain.HeaderCount() > 0 {
 			return

--- a/ledger/chainsync.go
+++ b/ledger/chainsync.go
@@ -935,12 +935,17 @@ func (ls *LedgerState) nextBufferedHeaderConnId() (
 func (ls *LedgerState) replayBufferedHeadersAsync(
 	connId ouroboros.ConnectionId,
 ) {
-	// Bail out if Close has already started so we never spawn a worker
-	// that issues DB reads after phase 3 closes the database (#2107).
+	// Hold replayMu so Close cannot start Wait between our closed check
+	// and Add(1); otherwise Wait could observe a zero counter and the
+	// subsequent Add(1) would panic with "WaitGroup misuse: Add called
+	// concurrently with Wait" (#2107).
+	ls.replayMu.Lock()
 	if ls.closed.Load() {
+		ls.replayMu.Unlock()
 		return
 	}
 	ls.replayWG.Add(1)
+	ls.replayMu.Unlock()
 	go func() {
 		defer ls.replayWG.Done()
 		ls.chainsyncMutex.Lock()

--- a/ledger/chainsync_replay_shutdown_test.go
+++ b/ledger/chainsync_replay_shutdown_test.go
@@ -1,0 +1,86 @@
+// Copyright 2026 Blink Labs Software
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ledger
+
+import (
+	"testing"
+	"time"
+)
+
+// Regression for #2107: replayBufferedHeadersAsync must be a no-op once
+// Close has been observed, so its goroutine can never reach DB reads
+// after the database is closed.
+func TestReplayBufferedHeadersAsyncSkippedAfterClose(t *testing.T) {
+	fixture := newChainsyncRollbackFixture(t)
+	ls := fixture.ls
+
+	// Simulate that Close has begun (closed=true) before the replay
+	// is scheduled. The async helper must not spawn a worker.
+	ls.closed.Store(true)
+
+	ls.replayBufferedHeadersAsync(fixture.connId)
+
+	done := make(chan struct{})
+	go func() {
+		ls.replayWG.Wait()
+		close(done)
+	}()
+	select {
+	case <-done:
+	case <-time.After(2 * time.Second):
+		t.Fatal("replayWG did not complete: a worker was spawned despite ls.closed=true")
+	}
+}
+
+// Regression for #2107: Close must drain in-flight replay goroutines
+// before returning, so callers (Node.shutdown phase 3) can safely close
+// the database without racing the replay's DB reads.
+func TestCloseWaitsForInFlightReplay(t *testing.T) {
+	fixture := newChainsyncRollbackFixture(t)
+	ls := fixture.ls
+
+	// Hold the chainsync mutex so the goroutine spawned by
+	// replayBufferedHeadersAsync blocks before touching the DB.
+	ls.chainsyncMutex.Lock()
+	ls.replayBufferedHeadersAsync(fixture.connId)
+
+	// Confirm a worker really did start and is now waiting on the lock.
+	// (The goroutine ran replayWG.Add(1) synchronously before launching.)
+	closeReturned := make(chan error, 1)
+	go func() {
+		closeReturned <- ls.Close()
+	}()
+
+	// Close must not return while the worker is still in flight.
+	select {
+	case err := <-closeReturned:
+		ls.chainsyncMutex.Unlock()
+		t.Fatalf("Close returned (err=%v) before replay drained", err)
+	case <-time.After(200 * time.Millisecond):
+	}
+
+	// Releasing the mutex lets the worker observe ls.closed=true and exit
+	// without issuing any DB reads.
+	ls.chainsyncMutex.Unlock()
+
+	select {
+	case err := <-closeReturned:
+		if err != nil {
+			t.Fatalf("Close returned error: %v", err)
+		}
+	case <-time.After(15 * time.Second):
+		t.Fatal("Close did not return after replay worker exited")
+	}
+}

--- a/ledger/chainsync_replay_shutdown_test.go
+++ b/ledger/chainsync_replay_shutdown_test.go
@@ -56,23 +56,36 @@ func TestCloseWaitsForInFlightReplay(t *testing.T) {
 	ls.chainsyncMutex.Lock()
 	ls.replayBufferedHeadersAsync(fixture.connId)
 
-	// Confirm a worker really did start and is now waiting on the lock.
-	// (The goroutine ran replayWG.Add(1) synchronously before launching.)
+	// (The goroutine ran replayWG.Add(1) synchronously before launching,
+	// so the wait group counter is non-zero from this point.)
 	closeReturned := make(chan error, 1)
 	go func() {
 		closeReturned <- ls.Close()
 	}()
 
-	// Close must not return while the worker is still in flight.
+	// Wait until Close has set closed=true and is therefore committed
+	// to draining the replay worker before it can return.
+	deadline := time.Now().Add(2 * time.Second)
+	for !ls.closed.Load() {
+		if time.Now().After(deadline) {
+			ls.chainsyncMutex.Unlock()
+			t.Fatal("Close did not set ls.closed within 2s")
+		}
+		time.Sleep(time.Millisecond)
+	}
+
+	// With closed=true and the worker blocked on chainsyncMutex, Close
+	// must not return: doing so would close the DB while the worker is
+	// still alive (the original #2107 panic).
 	select {
 	case err := <-closeReturned:
 		ls.chainsyncMutex.Unlock()
 		t.Fatalf("Close returned (err=%v) before replay drained", err)
-	case <-time.After(200 * time.Millisecond):
+	default:
 	}
 
 	// Releasing the mutex lets the worker observe ls.closed=true and exit
-	// without issuing any DB reads.
+	// without issuing any DB reads; Close can then finish draining.
 	ls.chainsyncMutex.Unlock()
 
 	select {

--- a/ledger/state.go
+++ b/ledger/state.go
@@ -496,7 +496,10 @@ type LedgerState struct {
 	// closed.Load() and Add(1) in handleEventChainUpdate.
 	rollbackMu sync.Mutex
 	// rollbackWG tracks in-flight rollback event emission goroutines
-	rollbackWG        sync.WaitGroup
+	rollbackWG sync.WaitGroup
+	// replayWG tracks in-flight replayBufferedHeadersAsync goroutines so
+	// Close can drain them before the database is closed (issue #2107).
+	replayWG          sync.WaitGroup
 	validationEnabled bool
 	// Sync progress reporting (Fix 4)
 	syncProgressLastLog  time.Time     // last time we logged sync progress
@@ -985,6 +988,30 @@ func (ls *LedgerState) Close() error {
 		)
 	}
 	ls.rollbackMu.Unlock()
+
+	// Drain in-flight replayBufferedHeadersAsync goroutines so they
+	// finish issuing DB reads before the owner closes the database
+	// (#2107). The closed flag set above prevents new goroutines from
+	// being spawned, so this Wait is bounded.
+	ls.config.Logger.Info("waiting for in-flight header replay goroutines")
+	replayStart := time.Now()
+	replayDone := make(chan struct{})
+	go func() {
+		ls.replayWG.Wait()
+		close(replayDone)
+	}()
+	select {
+	case <-replayDone:
+		ls.config.Logger.Info(
+			"header replay goroutines finished",
+			"elapsed", time.Since(replayStart).Round(time.Millisecond),
+		)
+	case <-time.After(10 * time.Second):
+		ls.config.Logger.Warn(
+			"timed out waiting for header replay goroutines",
+			"elapsed", time.Since(replayStart).Round(time.Millisecond),
+		)
+	}
 
 	// Stop slot clock
 	if ls.slotClock != nil {

--- a/ledger/state.go
+++ b/ledger/state.go
@@ -497,6 +497,10 @@ type LedgerState struct {
 	rollbackMu sync.Mutex
 	// rollbackWG tracks in-flight rollback event emission goroutines
 	rollbackWG sync.WaitGroup
+	// replayMu serializes replayWG.Add with Close's replayWG.Wait to
+	// prevent Add-after-Wait panics from the TOCTOU race between
+	// closed.Load() and Add(1) in replayBufferedHeadersAsync (#2107).
+	replayMu sync.Mutex
 	// replayWG tracks in-flight replayBufferedHeadersAsync goroutines so
 	// Close can drain them before the database is closed (issue #2107).
 	replayWG          sync.WaitGroup
@@ -991,27 +995,21 @@ func (ls *LedgerState) Close() error {
 
 	// Drain in-flight replayBufferedHeadersAsync goroutines so they
 	// finish issuing DB reads before the owner closes the database
-	// (#2107). The closed flag set above prevents new goroutines from
-	// being spawned, so this Wait is bounded.
+	// (#2107). Hold replayMu so no new goroutine can Add(1) between our
+	// closed flag and this Wait. The closed flag set above prevents
+	// new goroutines from being spawned, so the Wait is bounded by the
+	// in-flight workers; we wait unconditionally because returning
+	// while replay goroutines are still issuing DB reads would
+	// reintroduce the panic this fix is meant to prevent.
 	ls.config.Logger.Info("waiting for in-flight header replay goroutines")
 	replayStart := time.Now()
-	replayDone := make(chan struct{})
-	go func() {
-		ls.replayWG.Wait()
-		close(replayDone)
-	}()
-	select {
-	case <-replayDone:
-		ls.config.Logger.Info(
-			"header replay goroutines finished",
-			"elapsed", time.Since(replayStart).Round(time.Millisecond),
-		)
-	case <-time.After(10 * time.Second):
-		ls.config.Logger.Warn(
-			"timed out waiting for header replay goroutines",
-			"elapsed", time.Since(replayStart).Round(time.Millisecond),
-		)
-	}
+	ls.replayMu.Lock()
+	ls.replayWG.Wait()
+	ls.replayMu.Unlock()
+	ls.config.Logger.Info(
+		"header replay goroutines finished",
+		"elapsed", time.Since(replayStart).Round(time.Millisecond),
+	)
 
 	// Stop slot clock
 	if ls.slotClock != nil {

